doc/what-patch.1
.TH WHAT\-PATCH "1" "Debian Utilities" "DEBIAN"
.SH NAME
what\-patch \- detect which patch system a Debian package uses

.SH SYNOPSIS
.B what\-patch\fR [\fIoptions\fR]

.SH DESCRIPTION
\fBwhat\-patch\fR examines the \fIdebian/rules\fR file to determine which patch
system the Debian package is using.
.PP
\fBwhat\-patch\fR should be run from the root directory of the Debian source
package.

.SH OPTIONS
Listed below are the command line options for \fBwhat\-patch\fR:
.TP
.BR \-h ", " \-\-help
Display a help message and exit.
.TP
.B \-v
Enable verbose mode.
This will include the listing of any files modified outside or the \fIdebian/\fR
directory and report any additional details about the patch system if
available.

.SH AUTHORS
\fBwhat\-patch\fR was written by Kees Cook <kees@ubuntu.com>,
Siegfried-A. Gevatter <rainct@ubuntu.com>, and Daniel Hahler
<ubuntu@thequod.de>, among others.
This manual page was written by Jonathan Patrick Davies <jpds@ubuntu.com>.
.PP
Both are released under the GNU General Public License, version 3 or later.

.SH SEE ALSO
The Ubuntu MOTU team has some documentation about patch systems at the Ubuntu
wiki: \fIhttps://wiki.ubuntu.com/PackagingGuide/PatchSystems\fR

.PP
.BR cdbs\-edit\-patch (1),
.BR dbs\-edit\-patch (1),
.BR dpatch\-edit\-patch (1)


debian/NEWS
devscripts (2.10.54) unstable; urgency=low

    manpage-alert now, as originally intended, prints out the full path to the
    binary which is lacking a manpage instead of simply the binary name.

    Before:
      No manual entry for somebinary

    After:
      No manual entry for /usr/bin/somebinary

 -- James Vega <jamessan@debian.org>  Thu, 06 Aug 2009 18:03:41 -0400

devscripts (2.10.24) unstable; urgency=low

    Starting with devscripts 2.10.21, debchange attempts to
    automatically determine whether an NMU changelog should be
    started, by comparing DEBEMAIL and DEBFULLNAME to the
    package's Maintainer and Uploaders fields.

    This behaviour may now be disabled using the DEBCHANGE_AUTO_NMU
    configuration file variable.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Mon, 07 Apr 2008 14:38:08 +0100

devscripts (2.10.20) unstable; urgency=low

    devscripts 2.10.19 modified debcommit's --release option to use the
    uncommitted changes (if any) in the changelog to derive the commit
    message.

    This behaviour is now optional and can be enabled using the -R /
    --release-use-changelog command line option or by setting the
    DEBCOMMIT_RELEASE_USE_CHANGELOG configuration option to "yes".

    The default is to use the previous (i.e. pre-2.10.19) behaviour.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Wed, 19 Mar 2008 20:31:20 +0000

devscripts (2.10.19) unstable; urgency=low

    In devscripts 2.10.17, dget was modified to unpack source packages by
    default.  The new -d switch or the DGET_UNPACK configuration file option
    may be used to indicate that the unpacking should not occur.

    As linda has been removed from both unstable and testing, support for
    it has been removed from debuild and cvs-debuild.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Thu, 13 Mar 2008 19:33:38 +0000


scripts/what-patch.sh
#!/bin/bash
#
# Copyright 2006-2008 (C) Kees Cook <kees@ubuntu.com>
# Modified by Siegfried-A. Gevatter <rainct@ubuntu.com>
# Modified by Daniel Hahler <ubuntu@thequod.de>
#
# ##################################################################
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 3
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# See file /usr/share/common-licenses/GPL for more details.
#
# ##################################################################
#
# By default only the name of the patch system is printed.  Verbose mode can be
# enabled with -v.

if [ "$1" = "-h" ] || [ "$1" = "--help" ]
then
	cat <<EOM
Usage: $0 [-v]

Run this inside the source directory of a Debian package and it will detect
the patch system that it uses.

 -v: Enable verbose mode:
     - Print a list of all those files outside the debian/ directory that have
       been modified (if any).
     - Report additional details about patch systems, if available.
EOM
	exit 0
fi

while [ ! -r debian/rules ];
do
	if [ "$PWD" = "/" ]; then
		echo "Can't find debian/rules."
		exit 1
	fi
	cd ..
done

VERBOSE=0
if [ "$1" = "-v" ]
then
	VERBOSE=1
fi

if [ "$VERBOSE" -gt 0 ]; then
	files=`lsdiff -z ../$(dpkg-parsechangelog | grep ^Source: | sed -e "s/^Source: //")_$(dpkg-parsechangelog | grep ^Version: | sed -e "s/^Version: //").diff.gz 2>/dev/null | grep -v 'debian/'`
	if [ -n "$files" ]
	then
		echo "Following files were modified outside of the debian/ directory:"
		echo "$files"
		echo "--------------------"
		echo
		echo -n "Patch System: "
	fi
fi

if fgrep -q quilt debian/source/format 2>/dev/null; then
	echo "quilt"
	exit 0
fi

# Do not change the output of existing checks by default, as there are build
# tools that rely on the exisitng output.  If changes in reporting is needed,
# please check the "VERBOSE" flag (see below for examples).  Feel free
# to add new patchsystem detection and reporting.
for filename in $(echo "debian/rules"; grep ^include debian/rules | fgrep -v '$(' | awk '{print $2}')
do
	fgrep patchsys.mk "$filename" | grep -q -v "^#" && {
		if [ "$VERBOSE" -eq 0 ]; then
			echo "cdbs"; exit 0;
		else
			echo "cdbs (patchsys.mk: see 'cdbs-edit-patch')"; exit 0;
		fi
	}
	fgrep quilt "$filename" | grep -q -v "^#" && { echo "quilt"; exit 0; }
	fgrep dbs-build.mk "$filename" | grep -q -v "^#" && {
		if [ "$VERBOSE" -eq 0 ]; then
			echo "dbs"; exit 0;
		else
			echo "dbs (see 'dbs-edit-patch')"; exit 0;
		fi
	}
	fgrep dpatch "$filename" | grep -q -v "^#" && {
		if [ "$VERBOSE" -eq 0 ]; then
			echo "dpatch"; exit 0;
		else
			echo "dpatch (see 'patch-edit-patch')"; exit 0;
		fi
	}
	fgrep '*.diff' "$filename" | grep -q -v "^#" && {
		if [ "$VERBOSE" -eq 0 ]; then
			echo "diff splash"; exit 0;
		else
			echo "diff splash (check debian/rules)"; exit 0;
		fi
	}
done
[ -d debian/patches ] || {
	if [ "$VERBOSE" -eq 0 ]; then
		echo "patchless?"; exit 0;
	else
		echo "patchless? (did not find debian/patches/)"; exit 0;
	fi
}
if [ "$VERBOSE" -eq 0 ]; then
	echo "unknown patch system"
else
	echo "unknown patch system (see debian/patches/ and debian/rules)"
fi
